Implement and Fix issues #63 and #69

vladimir-v-diaz opened the pull request:
This final attack for the demo shows that the primary rejects metadata that is signed with revoked keys.
Fix and implementation of issues #63 and #69.

TODO for this pull request:
* [x] Display a banner for UnsignedMetadataError
* [x] Polish the text explaining the attack (in the README)
* [x] Receive feedback on approach before polishing the code

vladimir-v-diaz committed:
Fix bug and add functions to show prevention of revoked keys

vladimir-v-diaz committed:
Fix bug where the incorrect repository was being backed up

vladimir-v-diaz committed:
Add final attack to README

awwad commented on README.md:
````diff
@@ -500,3 +500,54 @@ to assure. Devices that have not been compromised in such an attack, however,
+```
+
+If the Primary performs an update cycle once again, it would appear to be
+up-to-date.  The metadata that was signed by the revoked keys should not
````
should not -> is not

vladimir-v-diaz committed:
Display DEFENDED banner if BadSignatureError is raised

vladimir-v-diaz committed:
Remove commented lines of code

vladimir-v-diaz committed:
Add Python interpreter prompts to code snippets and mention display of DEFENDED banner

vladimir-v-diaz committed:
Remove all of the prefix parameters (don't need the flexibility for the demo)

vladimir-v-diaz committed:
Rename the functions for the attack

vladimir-v-diaz committed:
Edit replaced function names in README

vladimir-v-diaz commented:
I removed the prefix parameters (simpler without them, and we don't need the flexibility for the demo) and renamed the functions.

Hopefully the following workflow is easier to follow:
```
dd.revoke_compromised_keys()

dp.update_cycle()

dd.sign_with_compromised_keys_attack()

dp.update_cycle() # BadSignatureError raised

dd.undo_sign_with_compromised_keys_attack()
```

vladimir-v-diaz committed:
Minor edit to comment and banner message

vladimir-v-diaz committed:
Review + revise sign_with_compromised_keys_attack()

vladimir-v-diaz committed:
Review + revise undo_signed_with_compromised_keys_attack()

vladimir-v-diaz committed:
Minor edit to text of final attack added to the README

vladimir-v-diaz committed:
Update README.md

Change word.

vladimir-v-diaz commented:
@awwad I finished revising the code and text.  You may now continue a full review of the PR

awwad commented on demo/demo_director.py:
```diff
@@ -190,16 +193,21 @@ def write_to_live(vin_to_update=None):
+      # Copy the staged metadata to a backup directory, which we'll move back
+      # into place when we later undo sign_with_compromised_keys_attack().
+      shutil.copytree(os.path.join(repo_dir, 'metadata.staged'),
+          os.path.join(repo_dir, 'metadata.backup'))
```
I think this is unrelated functionality that should live in its own function, run in the attack. It's occurring effectively at the end of write_to_live per repo. The new function, say backup_repositories(), can just loop over the repositories and back them up with that call. I'll add if you want.

vladimir-v-diaz replied:
Sure.  I will work to reproduce the unexpected ReplayedMetadataError and determine the cause and fix.

awwad commented on demo/demo_director.py:
```diff
@@ -259,17 +277,191 @@ def revoke_and_add_new_keys_and_write_to_live():
+  # Push the changes to "live", and make sure to back up the changes so that
+  # they can be restored later (e.g., after launching a
+  # sign_with_compromised_keys_attack()).
+  write_to_live(backup_repo=True)
```
sign_with_compromised_keys_attack should instead perform the backup, in a call separate from write_to_live

awwad commented on demo/demo_director.py:
```diff
@@ -259,17 +277,191 @@ def revoke_and_add_new_keys_and_write_to_live():
+  current_snapshot_private_key = director_service_instance.key_dirsnap_pri
+
+  # Set the new private keys in the director service.  These keys are shared
+  # between all vehicle repositories.
```
Copied comment somewhat misleading now. Maybe "Load the now-revoked keys" above and "tell Director service to use the old keys" here?

awwad commented on demo/demo_director.py:
```diff
@@ -259,17 +277,191 @@ def revoke_and_add_new_keys_and_write_to_live():
+    repository.timestamp.unload_signing_key(current_timestamp_private_key)
+
+    # Load the new signing keys to write metadata. The root key is unchanged,
+    # and in the demo it is already loaded.
```
new -> old, or something along those lines.

awwad commented on demo/demo_director.py:
```diff
@@ -259,17 +277,191 @@ def revoke_and_add_new_keys_and_write_to_live():
+  """
+  <Purpose>
+    Undo the actions executed by sign_with_compromised_keys_attack().  Namely,
+    move the valid metadata into the live and metadat.staged directories, and
```
typo: metadata

awwad commented on demo/demo_primary.py:
```diff
@@ -329,19 +329,25 @@ def update_cycle():
         if isinstance(exception.mirror_errors[mirror_url], tuf.ReplayedMetadataError):
           print_banner(BANNER_REPLAY, color=WHITE+BLACK_BG,
               text='The Director has instructed us to download a Timestamp'
-              ' that is older than the currently trusted version. This '
```
While it is likely, it is not necessary that this be the Timestamp, so I guess I should update that to "metadata".

vladimir-v-diaz replied:
But we specifically check for the Timestamp file on line 337.  This banner should never display for non-Timestamp metadata.

awwad replied:
Ah, you're right. (:

awwad committed:
DOC: Clarify result of compromise of both repositories

awwad reviewed (commented):
I'm getting a replayed metadata error after the last attack is undone, when I expect a successful update.

awwad committed:
Correct a missing reassignment after reloading repository

Previously, code did not save changes made to the existing
repository object.

awwad committed:
Rename revocation function in image repo to match director's

revoke_and_add_new_keys_and_write_to_live becomes
revoke_compromised_keys

vladimir-v-diaz committed:
Fix 'metadat.staged' typo

vladimir-v-diaz committed:
Merge branch 'attack_after_key_revocation' of github.com:vladimir-v-diaz/uptane into attack_after_key_revocation

vladimir-v-diaz committed:
Address 'new -> old' review comment

vladimir-v-diaz committed:
Fix misleading comment (review comment)

awwad committed:
Correct color of third defended banner

awwad committed:
Pull repository backup into its own function in the demo Director

and remove it from write_to_live

awwad committed:
Pull restore repositories code into its own function

awwad committed:
Merge branch 'attack_after_key_revocation' of https://github.com/vladimir-v-diaz/uptane into attack_after_key_revocation

awwad commented:
Okay, LGTM now. I made a few changes and tried the whole thing again. Take a quick peek at the changes and either of us can merge it.

vladimir-v-diaz commented on demo/demo_director.py:
```diff
@@ -193,11 +189,47 @@ def write_to_live(vin_to_update=None, backup_repo=False):
+    None.
+  """
+  # For each vehicle repository:
+  #   - write metadata.staged
```
Is this comment correct?  I don't think metadata.staged is written.

vladimir-v-diaz commented on demo/demo_director.py:
```diff
@@ -193,11 +189,47 @@ def write_to_live(vin_to_update=None, backup_repo=False):
+def backup_repositories():
+  """
+  <Purpose>
+    Back up the last-written state (contents of the 'metadata' directories in
```
Wouldn't _contents of the 'metadata.staged' directories_ be more accurate?

vladimir-v-diaz commented on demo/demo_director.py:
```diff
@@ -233,6 +233,80 @@ def backup_repositories():
+    # sign_with_compromised_keys_attack() was called.
+
+    if not os.path.exists(os.path.join(repo_dir, 'metadata.backup')):
+      import ipdb; ipdb.set_trace()
```
We should remove this import statement.

vladimir-v-diaz commented on demo/demo_director.py:
```diff
@@ -233,6 +233,80 @@ def backup_repositories():
+
+    # Copy the staged metadata to a temp directory, which we'll move into place
+    # atomically in a moment.
+    shutil.copytree(os.path.join(repo_dir, 'metadata.staged'),
```
Having print/log statements when deleting the `metadata` directories might also be useful (similar to the print statements above).  However, this isn't a blocker for the PR.

awwad committed:
Addressing review comments from Vlad

- Add some print statements to the backup and restore process
- Remove accidental ipdb import statement
- Correct comment referring to metadata instead of metadata.staged

awwad commented:
All good comments. Addressed. (:

vladimir-v-diaz commented:
PR looks good to me.  You can go ahead and merge!

xavo95 commented:
On 17th April you put:
- "I'm getting a replayed metadata error after the last attack is undone, when I expect a successful update."

I'm now facing the same issue with lastest commit, can you confirm? If after attack 3.7 you try to add a new target, the version of timestamp(and maybe other roles), roll back to 2, thus the primary is finding Replay Attack. Which is logic as the previous version was higher than 2.

awwad commented:
I have some revisions to the demo instructions to make. What version of the code are you using? (If you can just give me the commit, that'd work best.)

awwad commented:
I can't seem to reproduce the problem you're reporting. The attack and undo attack code in the demonstration are sensitive to state (for example, they back up and restore old metadata files to simulate replay attacks), so it's possible that if they're run out of order, the attacks might not work properly.

If you can provide more information (the code/commit you're using for example, and what you're doing), that would be appreciated.

xavo95 commented:
Replay attack, undo replay, compromise, uncompromise(add image on director
and image repo) and revoke keys, use old keys attack and recover. At this
point everything breaks. Also if you try to redo old keys attack it gets
verification key not found
El dj, 9 nov 2017 a les 20:15 Sebastien Awwad <notifications@github.com> va
escriure:

> I can't seem to reproduce the problem you're reporting. The attack and
> undo attack code in the demonstration are sensitive to state (for example,
> they back up and restore old metadata files to simulate replay attacks), so
> it's possible that if they're run out of order, the attacks might not work
> properly.
>
> If you can provide more information (the code/commit you're using for
> example, and what you're doing), that would be appreciated.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/uptane/uptane/pull/73#issuecomment-343261088>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AGbdir_aW8HqF5rQwzOgGRy26l4rx65pks5s009qgaJpZM4M9LwD>
> .
>

awwad commented:
I'm not seeing this problem. Can you tell me what commands you're running? (And this is the current develop branch code, I guess?)

Any error messages would be helpful, too.

xavo95 commented:
I found out the issue:

- On line 283 of demo/demo_director.py
```python
director_service_instance.vehicle_repositories[vin] = rt.load_repository(repo_dir)
```
Thought if the function load_repository is only supplied by one parameter, the new repo name will be setup to 'default' for every repo(different vins)

This implies that code like this will break as now the repo name is not the same as before
```python
roleinfo = tuf.roledb.get_roleinfo(self.rolename, self.repository_name)
```
or 
```python
tuf.roledb.mark_dirty(roles, self.repository_name)
```

If this commands are used after recovering from the 3.7 attack, you can't write_to_live nor revoke_keys nor anything that implies using a search by name
